### PR TITLE
Connect4 ai

### PR DESCRIPTION
**Pull request by lbiv4**
Finally done with this PR - there is still plenty more testing that can be done, but I'm ready to merge what I have since I think I dealt with most of the bugs. I'll try to add comments around important pieces, but some key overall things to note:

- I added a testing structure. These had to go into the same file as the components they are testing (it was a real pain to import into a testfile in a separate directory outside of `src`). Feel free to copy, tweak, or comment if you have any suggestions.
- There are a few tweaks to helper methods/things in `core.rs` - be aware of this as you merge, because some things are necessary (and others may be obsolete depending on what you've done on your helper functions branch.
- There can likely be tweaks to the probability/best move generation for the AI. To be honest I won't know how effective it is until I test it in a game - it's hard to unit test complicated scenarios like an actual game. But if you have ideas on how the AI should think, let me know.

### CODE REVIEW

File: src/connect4/ai.rs
Comment (lbiv4) [inline src/connect4/ai.rs]: This is in for debugging I was doing. I would remove, but I'm likely going to need it again when I look at this more thoroughly.

File: src/connect4/ai.rs
Comment (lbiv4) [inline src/connect4/ai.rs:146]: `use` was weird...I used `super::*` to get access to things in the outer scope (i.e. things in `AI`, but sometimes you still needed to import things like `Board` even though I had `use connect4::core::Board` in the file scope as well.

File: src/connect4/core.rs
Comment (lbiv4) [inline src/connect4/core.rs:52]: I don't know if this is ideal syntax...I like being able to write `grid.x`, for example, but feels weird to have `pub` fields in side the struct. If you want to use getter/setting like `getX()` or `setX()`, let me know. Probably a semi-large rework tho.

File: src/connect4/core.rs
Comment (lbiv4) [inline src/connect4/core.rs:305]: I forgot - I also reorganized this helper method. Slightly simpler code and logic was easier to test.

File: src/connect4/core.rs
Comment (lbiv4) [inline src/connect4/core.rs]: NOTE TO SELF - just do `run_len.min(3)`

File: src/connect4/core.rs
Comment (lbiv4) [inline src/connect4/core.rs:353]: Had to make a separate check because `get_run_in_direction` can return values from 0-4, but only values corresponding to runs of length 1-4 are held.

File: src/connect4/core.rs
Comment (lbiv4) [inline src/connect4/core.rs:442]: This is an example of why it'd be like to have a `Team` struct with a team number and color - had to pass in a random `MyColor::White` to `insert` when all I care about is which team it is.

**Review by lbiv4 — COMMENTED** [review]
Added some more explanatory comments. Feel free to ping me on Slack if you have comments/concerns

File: src/connect4/ai.rs
Comment (mukhsia) [inline src/connect4/ai.rs]: I think can be simplified to `use connect4::core::{GridPosition, Board, GameState, BOARD_SIZE, MyColor};`

File: src/connect4/core.rs
Comment (mukhsia) [reply src/connect4/core.rs:52]: For convenience, using pub instead of adding getter is alright...

**Review by mukhsia — APPROVED** [review]
Looks ok besides the stuff you mentioned needs changing. I'll merge this into my branch after fixes/ it is merged to connect4 branch

**Comment by lbiv4** [comment]
Merging PR after tweaks

**Pull request merged**
